File: solutions/python/linked-list/add-two-numbers-ii.py
# LeetCode 445 — Add Two Numbers II (Medium)
# Category: Linked List · Approach: Reverse + add
# Time: O(m + n) | Space: O(m + n)
# Source: https://leetcode.com/problems/add-two-numbers-ii/

def add_two_numbers(l1: list[int], l2: list[int]) -> list[int]:
    a, b = l1[::-1], l2[::-1]
    out = []
    carry = 0
    i = 0
    while i < len(a) or i < len(b) or carry:
        total = carry
        if i < len(a):
            total += a[i]
        if i < len(b):
            total += b[i]
        out.append(total % 10)
        carry = total // 10
        i += 1
    return out[::-1]

File: solutions/typescript/linked-list/add-two-numbers-ii.ts
// LeetCode 445 — Add Two Numbers II (Medium)
// Category: Linked List · Approach: Reverse + add
// Time: O(m + n) | Space: O(m + n)
// Source: https://leetcode.com/problems/add-two-numbers-ii/

function addTwoNumbers(l1: number[], l2: number[]): number[] {
  const a = l1.slice().reverse();
  const b = l2.slice().reverse();
  const out: number[] = [];
  let carry = 0;
  for (let i = 0; i < Math.max(a.length, b.length) || carry; i++) {
    const sum = (a[i] ?? 0) + (b[i] ?? 0) + carry;
    out.push(sum % 10);
    carry = Math.floor(sum / 10);
  }
  return out.reverse();
}
